boothj5/profanity
# Feature/omemo

Add native support to OMEMO

TODO:
- [x] OMEMO crypto materials generation
- [x] send OMEMO encrypted messages
- [x] receive OMEMO encrypted messages
- [x] long term storage for identity keys
- [x] add long term storage for session, ~~{signed,}prekeys,~~
- [x] get rid of libsodium
- [x] handle carbons
- [x] handle device_list_request response
- [x] check if prekey can be avoid while sending message
- [x] trust key from command line
- [x] untrust key from command line
- [x] valgrind check
- [ ] ~~add support for old gcm tag format parsing~~ (not being done for now since we are not aware of any client using it)
- [x] put gcm tag in key element (new format)
- [ ] ~~add support for signal pre v3 encryption~~ (not being done for now since we are not aware of any client using it)
- [x] start session on reception of a message with prekey (and republish bundle)
- [x] add support for libsignal 2.3.2
- [x] muc support
- [ ] ensure MUC has the right options (nonanonymous and access to fulljid)
- [x] Ensure PEP with bundle and device list are open (see [publish option](https://gist.github.com/iNPUTmice/7c52785ed69787516abb60e31703dbd2) )
- [ ] Check if `publish-options` is supported by server before attaching such node to query
- [x] Handle `precondition-not-met` error for `publish-options`
- [x] add autocompletion trust command
- [ ] add autocompletion untrust command
- [x] add a way to list trusted keys
- [ ] alert if sending message in a MUC with untrusted devices
- [ ] show list of known and trusted fingerprint on `/omemo start`
- [ ] add visual feedback when message has been sent on MUC

To test this PR you must install:

- gcrypt
- libsignal-protocol-c (>= v2.3.1)

On first run, ensure to create OMEMO cryptographic materials:
```
/omemo gen
```

To start an omemo session you must start it twice (issue about handling device_list_request response):
```
/omemo start buddy@example.org
```
Then you must trust some fingerprint:
```
/omemo fingerprint
/omemo trust <fingerprint>
```
